psiminer-core/src/main/kotlin/labelextractor/LabelExtractor.kt

```diff
@@ -4,22 +4,22 @@ import GranularityLevel
 import com.intellij.psi.PsiElement
 import psi.language.LanguageHandler
 
-data class LabeledTree(val root: PsiElement, val label: String)
+data class LabeledTree(val root: PsiElement, val label: Any?)
```

Comment:
That's the most tricky part of extending `PSIMiner` for using generic labels. Since this class stores information about labels, it should know about type. On the other side, all usages must have access to direct representation without explicit details. For example, in storage, you add cast to String, but there is no guarantee that this cast would be successful.

What I suggest is to parameterize this class with generic type of label and add property to work with string representation:
```kotlin
open class LabeledTree<T>(val root: PsiElement, private val fullLabel: T?) {
    open val label: String
        get() = label.toString()
}
```

Therefore, all storages will work only with `label` without knowing details about a label.
Also, you may define your own derived class for `LabeledTree` to change the behavior of converting `label` to string. For example:
```kotlin
class LabeledWithParametersTree(root: PsiElement, private val parameters: List<String>) :
    LabeledTree<List<String>>(root, parameters) {
    override val labelRepresentation: String
        get() = parameters.joinToString(",")
}
```

Reply:
This idea is interesting. But a list labels is joined and returned as String. We could do this in the implementation of handleTree itself. But again that could create code duplicate if are trying to do the same again in other implementations. I am looking on this, I will let you know if I get any idea also.

Reply:
Converting to String inside `handleTree` isn't a good idea, because it will reduce the flexibility of processing trees. For example, if you in the future would need to filter some types from these parameters or change format then you need to dig into this method again. What I suggest, you store parameters inside the special class, and therefore, all logic about working with labels in the future will be encapsulated inside a small method.

Reply:
I am not quite sure about creating the LabelWithParameterTree. Does that mean we have to check the type of instance returned by handleTree method and create the instance of LabeledTree?